### PR TITLE
(Closes #2138) add support for expressions in array bound definitions

### LINKED ISSUE #2138: Add support to parse Fortran array bounds with expressions
For example: `real :: a(start - 1: stop + 1)`

Currently it is a UnknownFortranType, but it should be an ArrayType with an ArrayBound with 2 psyir expressions.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Integration tests failed due to attempt to call `walk` on an `Extent` at line 293 of `sympy_writer.py`.

**Comment by arporter** [comment]
The latest NEMO code ends up with some pretty hairy array indexing and declarations due to the tiling. I've added an example to one of the tests. Integration tests are re-running (it was only the NEMO ones that failed).

**Comment by arporter** [comment]
Integration tests were green second time around. I just need to address missing coverage.

**Comment by arporter** [comment]
Ready for review now. A fairly small change but ensuring that it worked when handling the declaration of members of a structure was quite difficult.
One for @LonelyCat124, @hiker or @sergisiso.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 1]
```diff
@@ -4486,7 +4466,16 @@ def _contains_intrinsic_reduction(pnodes):
...
+                lbound = mask_shape[idx-1].lower
+                if isinstance(lbound, Literal) and lbound.value == "1":
+                    # Lower bound is unity so size is just the upper bound.
+                    expr2 = mask_shape[idx-1].upper.copy()
+                else:
+                    # Size = upper-bound - lower-bound + 1
+                    expr = BinaryOperation.create(
+                        sub_op, mask_shape[idx-1].upper.copy(), lbound.copy())
+                    expr2 = BinaryOperation.create(add_op, expr, one.copy())
+                loop.addchild(expr2)
```
Comment (sergisiso) [inline src/psyclone/psyir/frontend/fparser2.py:4478]: Nice improvement but I am unsure if this covers the case where there is a non-unity lower but no upper, like when remapping a argument lowerbound:
```
subroutine test(a)
   integer, dimension(5:) :: a
```
Could you add a test?

**Review by sergisiso — CHANGES_REQUESTED** [review]
Well done @arporter this PR improves many things in addition to bound expressions, like connecting symbols on derived types and improve where support and I like how it is implemented. There is just minor comments.

File: src/psyclone/tests/psyir/nodes/array_mixin_test.py [hunk 2]
```diff
@@ -153,17 +153,29 @@ def test_is_upper_lower_bound(fortran_reader):
...
+      trdt(ntsi-(0):,ntsj-(0):ntej+(0),:) =    &
+         ptrd(ntsi-(0):ntei+(0),ntsj-(0):ntej+(0),:)
+      a(1:n) = 0.0
+    end subroutine
+    '''
     psyir = fortran_reader.psyir_from_source(code)
     assigns = psyir.walk(Assignment)
-    array_ref = assigns[0].lhs
+    trdt_ref = assigns[0].lhs
+    assert trdt_ref.is_upper_bound(0)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/array_mixin_test.py:170]: Yep, done.

File: src/psyclone/tests/psyir/nodes/array_mixin_test.py [hunk 3]
```diff
@@ -153,17 +153,29 @@ def test_is_upper_lower_bound(fortran_reader):
...
+    assert not trdt_ref.is_upper_bound(1)
+    assert trdt_ref.is_upper_bound(2)
+    ptrd_ref = assigns[0].rhs
+    assert not ptrd_ref.is_upper_bound(0)
+    assert ptrd_ref.is_upper_bound(2)
+    # Return True as the symbolic values of the declaration and array
+    # reference match.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/nodes/array_mixin_test.py:177]: Good call - done.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 4]
```diff
@@ -1311,81 +1326,25 @@ def _parse_dimensions(dimensions, symbol_table):
...
             represents a scalar.
-        :rtype: list of NoneType or 2-tuples of \
-                :py:class:`psyclone.psyir.nodes.DataNode`
+        :rtype: list[Optional[
+            tuple[:py:class:`psyclone.psyir.nodes.DataNode`,
+                  :py:class:`psyclone.psyir.nodes.DataNode`]]]
 
-        :raises NotImplementedError: if anything other than scalar, integer \
+        :raises NotImplementedError: if anything other than scalar, integer
             literals or symbols are encounted in the dimensions list.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:1345]: Oops, good spot.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 5]
```diff
@@ -1311,81 +1326,25 @@ def _parse_dimensions(dimensions, symbol_table):
         UnresolvedType.
 
         :param dimensions: fparser dimension attribute.
-        :type dimensions: \
+        :type dimensions:
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:1329]: Oops.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 6]
```diff
@@ -1887,6 +1845,8 @@ def _process_decln(self, scope, symbol_table, decl, visibility_map=None,
         '''
         # pylint: disable=too-many-arguments
         (type_spec, attr_specs, entities) = decl.items
+        # Whether or not we are dealing with a member of a Structure.
+        is_member = isinstance(decl, Fortran2003.Data_Component_Def_Stmt)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:1849]: Agreed. Gone now and comment updated.

File: src/psyclone/psyir/frontend/fparser2.py [hunk 7]
```diff
@@ -2220,15 +2187,29 @@ def _process_derived_type_decln(self, parent, decl, visibility_map):
...
+            # Re-use the existing code for processing symbols. This needs to
+            # be able to find any symbols declared in an outer scope but
+            # referenced within the type definition (e.g. a type name). Hence
+            # the table needs to be connected to the tree.
+            local_table = Container("tmp", parent=parent).symbol_table
+            local_table.default_visibility = default_compt_visibility
+
             for child in walk(decl, Fortran2003.Data_Component_Def_Stmt):
                 self._process_decln(parent, local_table, child)
             # Convert from Symbols to type information
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:2199]: Yes, that's much better.

File: src/psyclone/psyir/frontend/fparser2.py (hunk 1 above)
Comment (arporter) [reply src/psyclone/psyir/frontend/fparser2.py:4478]: I've done my best. If it's not what you had in mind then please say :-)

**Comment by arporter** [comment]
There was an indirect coverage change, hence the new test. Ready for another look now.

**Review by sergisiso — APPROVED** [review]
@arporter All comments have been addressed, the CI, integration and docs builing all succeeds. This is ready to merge.

**Pull request merged**